DRA API: implement ResourceClaim strategy for DRADeviceTaints

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Dropping the disabled "Tolerations" field in the ResourceClaim API was missing.

This wasn't possible at the time of implementing the Device Taints API, at least not completely, because it depended on prioritized list being merged first, to cover the "FirstAvailable" field introduced together with that feature.

That the device taints PR got merged despite this gap was an oversight. The confusing TODO probably didn't help: the entire implementation was missing (or got lost due to a bad merge conflict resolution, not sure anymore) and it referenced the wrong other feature (partitionable devices doesn't affect ResourceClaim).

#### Which issue(s) this PR is related to:

KEP: https://github.com/kubernetes/enhancements/issues/5055

#### Special notes for your reviewer:

This allowed clients to set the field when it should have been dropped. It simply had no effect. Clients can keep updating such objects because of the "feature in use" check, as long as the spec remains immutable.

#### Does this PR introduce a user-facing change?

```release-note
DRA API: the "tolerations" field in exact and sub requests now gets dropped properly when the DRADeviceTaints API is disabled.
```

pohly committed:
DRA API: implement ResourceClaim and ResourceClaimTemplate strategy for DRADeviceTaints

This wasn't possible at the time of implementing the Device Taints API, at
least not completely, because it depended on prioritized list being merged
first, to cover the "FirstAvailable" field introduced together with that
feature.

That the device taints PR got merged despite this gap was an oversight. The
confusing TODO probably didn't help: the entire implementation was missing (or
got lost due to a bad merge conflict resolution, not sure anymore) and it
referenced the wrong other feature (partitionable devices doesn't affect
ResourceClaim).

For some reason, ResourceClaimTemplate update testing was less complete than
the update testing of ResourceClaim. Fixed by copying the entire
TestStrategyUpdate over and switching it to testing ResourceClaimTemplates.

pohly commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -212,6 +212,7 @@ func toSelectableFields(claim *resource.ResourceClaim) fields.Set {
 // dropDisabledFields removes fields which are covered by a feature gate.
 func dropDisabledFields(newClaim, oldClaim *resource.ResourceClaim) {
 	dropDisabledDRAPrioritizedListFields(newClaim, oldClaim)
+	dropDisabledDRADeviceTaintsFields(newClaim, oldClaim) // Intentionally after dropDisabledDRAPrioritizedListFields to avoid iterating over FirstAvailable slice which needs to be dropped.
```
Yes. Will add it there, too. Good catch!

pohly commented on pkg/registry/resource/resourceclaim/strategy_test.go:
```diff
@@ -372,6 +419,26 @@ func TestStrategyCreate(t *testing.T) {
 				}
 			},
 		},
+		"drop-fields-device-taints": {
```
Yes.

pohly commented on pkg/registry/resource/resourceclaim/strategy_test.go:
```diff
@@ -587,6 +656,54 @@ func TestStrategyUpdate(t *testing.T) {
 				}
 			},
 		},
+		"drop-fields-device-taints-in-prioritized-list": {
```
Yes. Now I am questioning my own sanity. I could have sworn that I had added them.

pohly commented:
/assign @mortent 

Please check again, it should be complete now.

liggitt commented:
> Shall we still try to get this included in v1.34? It's "only" for an alpha feature, but getting the API right is important.
> 
> /assign @liggitt

I just saw this, paging in context ... this is fixing clearing an alpha field which was added in 1.33 in https://github.com/kubernetes/kubernetes/pull/130447?

I think it's too late for any non-release-blocking changes in 1.34, and if 1.33 already released without this, I don't think we'd consider it release-blocking for 1.34 either.

pohly commented:
> this is fixing clearing an alpha field which was added in 1.33 in https://github.com/kubernetes/kubernetes/pull/130447?

Correct.

> milestone v1.35

Ack, makes sense.

tallclair commented on pkg/registry/resource/resourceclaimtemplate/strategy.go:
```diff
@@ -113,10 +113,46 @@ func toSelectableFields(template *resource.ResourceClaimTemplate) fields.Set {
 
 func dropDisabledFields(newClaimTemplate, oldClaimTemplate *resource.ResourceClaimTemplate) {
```
Should we dedupe the logic between this and the ResourceClaim strategy? I think it just needs a `func dropDisabledFields(newResourceClaimSpce, oldResourceClaimSpec *resource.ResourceClaimSpec)`?

pohly committed:
DRA API: unify dropped field logic for ResourceClaim and ResourceClaimTemplate

The spec is the same in both, so those fields are now handled by common
code. For ResourceClaim spec updates, the "in use" check now only considers
the spec.

Theoretically some features could be in use in an old ResourceClaim status
and not in use in the spec. This can only occur in a spec update, which is
currently prevented because the entire spec is immutable. Even if it was
allowed, preventing adding disabled fields to the spec is the right thing to
do regardless of what may have ended up in the status earlier.

pohly commented:
/test pull-kubernetes-node-e2e-containerd-2-0-dra-alpha-beta-features

We had a test flake with some tests not cleaning up properly after themselves, should be fixed now (
https://github.com/kubernetes/kubernetes/pull/134047).

liggitt commented:
Thanks for the update, @tallclair will take a pass and then we can get this in

tallclair commented on pkg/registry/resource/resourceclaimspec/strategy.go:
```diff
@@ -0,0 +1,177 @@
+limitations under the License.
+*/
+
+package resourceclaimspec
```
It looks a little strange to have a registry directory for something that isn't a top-level resource. It looks like every other instance of an exported `DropDisabled.*Fields` function is under `pkg/api/.../util.go` (e.g. `pkg/api/resourceclaim(spec)/util.go`)

tallclair reviewed (commented):
lgtm, just a nit about the file path.
